### DISCUSSION TITLE
Understanding typing collections.

### DISCUSSION THREAD

**Discussion by iBotPeaches** [larastan/larastan#1185] [Q&A]
I just got done with a Laravel 9 upgrade + jumping to v2 of Larastan. Collections all over are failing with the same warning and I'm struggling to resolve.

The Model
```php
// (class docblock)
/**
 * @property-read Collection<int,PullRequest> $pullRequests
 */
 
// (relation)
public function pullRequests(): HasMany
{
    return $this->hasMany(PullRequest::class);
}
 ```
 
 The affected code.
```php
// $repository is the above model
$repository->pullRequests->each(function (PullRequest $pullRequest) { };
```

I get:

```
Parameter #1 $callback of method Illuminate\Support\Collection<int,App\Models\PullRequest>::each() expects
  callable(int, int): mixed,
  Closure(App\Models\PullRequest): void given.  
```

I found a thread (https://github.com/nunomaduro/larastan/issues/1139) that was similar, but folks in there are wrapping collections at both the Collection level and Model level with `newCollection`. I'm just refusing to believe you have to wrap so many things in order for templates to properly apply, but want to be easily told its required if thats how it works.

Since I used to do `Collection|PullRequest[] $pullRequests` on the Model, which worked great for typing until this.

I then found this thread (https://github.com/nunomaduro/larastan/discussions/1141) which again pointed at being forced to wrap each entity collection with its own. Is that the only way? I don't really want to make a new collection class for every entity in the system.

What am I missing?

**Comment by canvural** [comment]
Hi,

Which version of Laravel are you using? There were some issue with Laravel 9.0.x which would lead to the errors you are getting. If you upgrade to for example 9.5 it should be fine.

Also the stuff about wrapping you mentioned should only apply if you have custom collection classes for your model. Which is not your case.

Also as a side note, you don't need to have that `property-read` tag. Larastan can correctly infer that already. If static analysis was the only reason to put it there.

**Reply by iBotPeaches** [reply]
Hi,

I'm using 9.5.1 presently of Laravel. The `property-read` I think started to support an IDE a few major versions ago and we just kept the development pattern up. With smarter IDEs, linting and more I can properly retire most of those.

**Reply by canvural** [reply]
I'm sorry but it really should work if you have Laravel 9.5.1 I cannot reproduce this.

If possible try to create a small repo that reproduces this issue. Otherwise double check the Laravel version or try to clear the PHPStan cache.

**Reply by iBotPeaches** [reply]
Sounds good. Thanks for the help. I'll dig into it more and look into a new test repo which may help me spot an issue if it works out.

**Reply by iBotPeaches** [reply]
New repo has no issue with a similar example. Thanks, I will just redo everything and see where I messed up the existing large project.
